clang/lib/CConv/ConstraintVariables.cpp

```diff
@@ -395,10 +393,32 @@ void PointerVariableConstraint::dump_json(llvm::raw_ostream &O) const {
 
 void PointerVariableConstraint::getQualString(uint32_t TypeIdx,
                                               std::ostringstream &Ss) {
-  std::map<ConstraintKey, Qualification>::iterator Q = QualMap.find(TypeIdx);
-  if (Q != QualMap.end())
-    if (Q->second == ConstQualification)
-      Ss << "const ";
+  auto QIter = QualMap.find(TypeIdx);
+  if (QIter != QualMap.end()) {
+    for (Qualification Q : QIter->second) {
+      switch (Q) {
+      case ConstQualification:
+        Ss << "const ";
+        break;
+      case VolatileQualification:
+        Ss << "volatile ";
+        break;
+      case RestrictQualification:
+        Ss << "restrict ";
+        break;
+      }
+    }
+  }
+}
+
+void PointerVariableConstraint::insertQualType(uint32_t TypeIdx,
+                                               QualType &QTy) {
+  if (QTy.isConstQualified())
+    QualMap[TypeIdx].insert(ConstQualification);
+  if (QTy.isVolatileQualified())
+    QualMap[TypeIdx].insert(VolatileQualification);
+  if (QTy.isRestrictQualified())
+    QualMap[TypeIdx].insert(RestrictQualification);
```

Comment:
Is there any way we preserve the order in which these qualifiers are declared? 
I am thinking about the case where we have `restrict volatile` and we end up rewriting `volatile restrict`. I am not so adept with C type qualifiers. So I am not sure if this matters or now.

Reply:
I'm fairly sure the order of qualifiers doesn't matter.

Reply:
I agree, it should matter. The only tricky thing is that `const int *` is not the same as `int * const`; but that has to do with position, not order (in the same position).

Reply:
From the [C11 spec](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1570.pdf):
![image](https://user-images.githubusercontent.com/1738940/87171569-d5751f00-c2a0-11ea-9535-e1542783b35e.png)